File: backend/src/main/java/com/assetflow/dashboard/DashboardService.java
package com.assetflow.dashboard;

import com.assetflow.asset.repository.AssetRepository;
import com.assetflow.assetcategory.repository.AssetCategoryRepository;
import com.assetflow.booking.repository.BookingRepository;
import com.assetflow.common.enums.AssetStatus;
import com.assetflow.common.enums.BookingStatus;
import com.assetflow.dashboard.dto.DashboardStats;
import com.assetflow.department.repository.DepartmentRepository;
import com.assetflow.employee.repository.EmployeeRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
@RequiredArgsConstructor
public class DashboardService {

    private final AssetRepository assetRepository;
    private final BookingRepository bookingRepository;
    private final EmployeeRepository employeeRepository;
    private final DepartmentRepository departmentRepository;
    private final AssetCategoryRepository assetCategoryRepository;

    @Transactional(readOnly = true)
    public DashboardStats getStats() {
        long totalAssets = assetRepository.count();

        // Asset counts by status — use repository count queries
        long available = assetRepository.countByStatus(AssetStatus.AVAILABLE);
        long allocated = assetRepository.countByStatus(AssetStatus.ALLOCATED);
        long underMaintenance = assetRepository.countByStatus(AssetStatus.UNDER_MAINTENANCE);
        long disposed = assetRepository.countByStatus(AssetStatus.DISPOSED);

        // Booking counts
        long activeBookings = bookingRepository.countByStatusIn(
                List.of(BookingStatus.APPROVED));
        long pendingBookings = bookingRepository.countByStatusIn(
                List.of(BookingStatus.PENDING));

        return DashboardStats.builder()
                .totalAssets(totalAssets)
                .availableAssets(available)
                .allocatedAssets(allocated)
                .underMaintenanceAssets(underMaintenance)
                .disposedAssets(disposed)
                .activeBookings(activeBookings)
                .pendingBookings(pendingBookings)
                .totalEmployees(employeeRepository.count())
                .totalDepartments(departmentRepository.count())
                .totalCategories(assetCategoryRepository.count())
                .build();
    }
}

File: backend/src/dashboard/dashboard.service.ts
import { Injectable } from '@nestjs/common';
import { InjectModel } from '@nestjs/mongoose';
import { Model } from 'mongoose';
import { Asset } from '../asset/schemas/asset.schema';
import { Booking } from '../booking/schemas/booking.schema';
import { Employee } from '../employee/schemas/employee.schema';
import { Department } from '../department/schemas/department.schema';
import { AssetCategory } from '../asset-category/schemas/asset-category.schema';
import { AssetStatus } from '../common/enums/asset-status.enum';
import { BookingStatus } from '../common/enums/booking-status.enum';

export interface DashboardStats {
  totalAssets: number;
  availableAssets: number;
  allocatedAssets: number;
  underMaintenanceAssets: number;
  disposedAssets: number;
  activeBookings: number;
  pendingBookings: number;
  totalEmployees: number;
  totalDepartments: number;
  totalCategories: number;
}

@Injectable()
export class DashboardService {
  constructor(
    @InjectModel(Asset.name) private assetModel: Model<any>,
    @InjectModel(Booking.name) private bookingModel: Model<any>,
    @InjectModel(Employee.name) private employeeModel: Model<any>,
    @InjectModel(Department.name) private departmentModel: Model<any>,
    @InjectModel(AssetCategory.name) private categoryModel: Model<any>,
  ) {}

  async getStats(): Promise<DashboardStats> {
    const [
      totalAssets,
      availableAssets,
      allocatedAssets,
      underMaintenanceAssets,
      disposedAssets,
      activeBookings,
      pendingBookings,
      totalEmployees,
      totalDepartments,
      totalCategories,
    ] = await Promise.all([
      this.assetModel.countDocuments(),
      this.assetModel.countDocuments({ status: AssetStatus.AVAILABLE }),
      this.assetModel.countDocuments({ status: AssetStatus.ALLOCATED }),
      this.assetModel.countDocuments({ status: AssetStatus.UNDER_MAINTENANCE }),
      this.assetModel.countDocuments({ status: AssetStatus.DISPOSED }),
      this.bookingModel.countDocuments({ status: BookingStatus.APPROVED }),
      this.bookingModel.countDocuments({ status: BookingStatus.PENDING }),
      this.employeeModel.countDocuments(),
      this.departmentModel.countDocuments(),
      this.categoryModel.countDocuments(),
    ]);

    return {
      totalAssets,
      availableAssets,
      allocatedAssets,
      underMaintenanceAssets,
      disposedAssets,
      activeBookings,
      pendingBookings,
      totalEmployees,
      totalDepartments,
      totalCategories,
    };
  }
}
